dislib/decomposition/tsqr/base.py

```diff
@@ -71,293 +67,693 @@ def tsqr(a: Array, n_reduction=2, mode="complete", indexes=None):
+                                      reg_shape=(a._reg_shape[1],
+                                                 a._reg_shape[1]),
+                                      shape=(
+                                          a.shape[1] if
+                                          a.shape[1] < a._reg_shape[0]
+                                          else a._reg_shape[0], a.shape[1]),
+                                      sparse=False)
+                else:
+                    small_q = Array(q_blocks,
+                                    top_left_shape=(a._reg_shape[1],
+                                                    a._reg_shape[1]),
+                                    reg_shape=(a._reg_shape[1],
+                                               a._reg_shape[1]),
+                                    shape=(a.shape[1] if
+                                           a.shape[1] < a._reg_shape[0]
+                                           else a._reg_shape[0],
+                                           a.shape[1]), sparse=False)
+
+                q_1 = matmul(qsaux.pop(0), small_q)
                 if len(qsaux) > 0:
-                    q_2 = matmul(qsaux.pop(0), small_q[a._reg_shape[1]:])
+                    q_2 = matmul(qsaux.pop(0), small_q_2)
                     qs.append(concat_rows(q_1, q_2))
                 else:
                     qs.append(q_1)
-                rs.append(r)
-        r_blocks = [[object()]]
-        _construct_blocks(r_blocks, r, (a.shape[1], a.shape[1]))
-        r = _empty_array(shape=(a.shape[1], a.shape[1]),
-                         block_size=(a.shape[1], a.shape[1]))
-        r._blocks = r_blocks
+                rs.append(r_blocks)
+                last_block_irregular = False
+        r = Array(rs[0],
+                  top_left_shape=(a._reg_shape[1], a._reg_shape[1]),
+                  reg_shape=(a._reg_shape[1], a._reg_shape[1]),
+                  shape=(a.shape[1], a.shape[1]), sparse=False)
         return qs[-1], r
+
     elif mode == "reduced_inverse":
-        if n_reduction == 2:
-            if _is_not_power_of_two(a._n_blocks[0]):
-                raise ValueError("This mode only works if the number of "
-                                 "blocks is a direct power of the "
-                                 "reduction number")
-        else:
-            if _is_not_power(n_reduction, a._n_blocks[0]):
-                raise ValueError(
-                    "This mode only works if the number of blocks is a"
-                    " direct power of the reduction number")
-        total_depth = 1
-        for block in a._blocks:
-            q, r = _compute_qr([block], "reduced")
-            qs.append(q)
-            rs.append(r)
+        if _is_not_power_of_two(a._n_blocks[0]):
+            raise ValueError("This mode only works if the number of "
+                             "blocks is a direct power of the "
+                             "reduction number")
```

Comment:
Same as previous comment: This mode only works if the number of blocks is a direct power of 2